clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -4022,8 +4169,36 @@ LValue CodeGenFunction::EmitArraySubscriptExpr(const ArraySubscriptExpr *E,
       ArrayLV = EmitArraySubscriptExpr(ASE, /*Accessed*/ true);
     else
       ArrayLV = EmitLValue(Array);
+
     auto *Idx = EmitIdxAfterBase(/*Promote*/true);
 
+    if (SanOpts.has(SanitizerKind::ArrayBounds)) {
```

Comment:
I'm honestly not sure how to do that. I may have missed a method, but if you have the base pointer how do you use that to compute the rest of the expression? For instance, if `p->p->p` is the base, how do I use that to emit the full expression `p->p->p->a.b.c.d.e[index]`?

I can skip the extra load here by using the pointer from the `ArrayLV` GEP.

Reply:
Okay, but that seems incredibly fragile. It'll be bypassing the code in `EmitLValue`. Wouldn't it be better to do it in the opposite way? Allow `EmitLValue` to do its thing. It'll return a GEP. I can then use `ArrayLV.getAddress(*this)` to build a GEP to `count` without re-loading the base pointer?

Reply:
So instead of a GEP from the base to "array", then a GEP from the base to "count", you emit a GEP from the base to "array", then a GEP from the "array" to the "count"?  That's basically just rearranging the math; instead of "a + b", you compute "a + c + (b - c)".  Maybe it saves refactoring some code, though.

Reply:
There's a key difference. Emitting the code to access "array" is potentially far more complex than the code to access "count". This is because "count" will be either at the top level of the structure or in an anonymous structure at the top level. The GEP indices are therefore much simpler to calculate.

Reply:
Responding to your comment about emitting the base twice, this seems to be, if not common, at least not unusual. Given the proximity of the array subscript expr and the counted by checks, any duplicate base pointer loads are elided.

Reply:
In general, emitting an expression twice can miscompile and/or crash the compiler.  I'm not concerned about performance optimizations.

Reply:
@rapidsna It would ignore that case, because it could have side-effects. But if it's the base pointer, it could be reused. I added a method in the visitor. I quickly tried the `CodeGenFunction::EmitLValueForField` method. Does it work with fields that aren't at the top level? At first glance, it doesn't appear to do that.

@efriedma-quic Okay. I changed it to look through the GEPs for the base pointer.

Reply:
Recursively skipping over all GEPs seems, at best, extremely fragile; there could be GEPs for reasons other than the member expressions you care about.

Reply:
Unfortunately, all pointers are opaque now, so there's no way to tell what type it is. So it's hard to know when to stop.

Reply:
Right. I've suggested a couple alternatives.  (You can explicitly compute the offset between the array and corresponding counted_by field.  Or you can start at the relevant base, and write out code to GEP to both the counted_by field and the array.)

Reply:
So just to be clear, you want me to bypass this code:

```
    LValue ArrayLV;
    // For simple multidimensional array indexing, set the 'accessed' flag for
    // better bounds-checking of the base expression.
    if (const auto *ASE = dyn_cast<ArraySubscriptExpr>(Array))
      ArrayLV = EmitArraySubscriptExpr(ASE, /*Accessed*/ true);
    else
      ArrayLV = EmitLValue(Array);
    auto *Idx = EmitIdxAfterBase(/*Promote*/true);
```

and instead calculate the base pointer from the `Array` that will point to both `count` and `array`. Then emit that pointer, and then append to that GEPs to access `count` and `array` that I calculate by hand?

That sounds incredibly fragile, given the complexity of both of the above `Emit*Expr` methods. But if that will finally get some approval, I'll cobble it together. It'll almost certainly fail for some side case, but I'm stupid, so what do I know.

Reply:
Yes.  You should be able to refactor some code out of CodeGenFunction::EmitMemberExpr so it's not a completely independent codepath.

I mean, I see why you're concerned about re-implementing EmitMemberExpr, but fragility isn't just about simplifying the control flow within the compiler; the nastiest problems are caused by one piece of code making assumptions about how an apparently unrelated part of the code behaves.

Reply:
> You should be able to refactor some code out of CodeGenFunction::EmitMemberExpr so it's not a completely independent codepath.

I could use it as a template, but refactoring probably won't work very well, because even if I could call it to generate the FAM array access, I don't have an Expr for the count.

> the nastiest problems are caused by one piece of code making assumptions about how an apparently unrelated part of the code behaves.

This is why I'm being very stubborn about *not* skipping the existing Emit functions. The assumption we're making are that generating the GEPs by hand for both fields is "easy". We can make that assumption for the `count` field, because of the expectations I mentioned above (how it's at the top-level of the struct, etc.). But I don't want to make those assumptions about the flexible array field.

As it turns out, I should be able to use `getSourceElementType` on a GEP to know when to stop skipping GEPs.

Keep in mind, the code we generate is for bounds checking, so we have a fallback if we're unable to generate a proper bounds check using the `count`. I.e., if we run into something unexpected, we can abort the bounds checking generation.

Reply:
Can we not just reverse the member access to get back to the root of the `counted_by` path?  Presumably the root of the `counted_by` path for a field is always the non-transparent context of the field.  This should always be at a constant offset from the address point of the field — there's no such thing as an anonymous array.  Or am I misunderstanding something important about the constraints on `counted_by` paths?

Reply:
Yeah.  Specifically, I was thinking that you already have the address of this flexible array member, and if the immediately containing struct is uniquely allocated directly within the root type of the `__counted_by` path, you can just find that offset statically and subtract to get a pointer to the root type.  But of course that only works if it *is* uniquely allocated directly within the root type, which it isn't going to be in the general case if the root type is just some lexically-enclosing struct, which ties this back to the other conversation.

Stepping back for a second, I'm surprised you need to do so much from scratch for flexible arrays.  Don't the other `-fbounds-safety` features already have some logic where they recognize e.g. a load from a `__counted_by` field of pointer type and then emit the count expression?  How do those code paths solve this problem?

Reply:
I have checks in place to make sure that I can find the correct enclosing type. For instance, if I have something like:

```
struct B {
  int count;
  int arr[] __counted_by(count);
};

struct A {
  int count;
  struct B *b;
};
```

when looking at `someA->b->arr[idx]`, it only cares about looking at the `count` in `struct B`. The `count` in `struct A` won't be seen. This is because we're looking for the most-enclosing struct of `arr`, which is `struct B`. Now if I have something like:

```
struct A {
  int count;
  struct B {
    int count;
    int arr[] __counted_by(count);
  } *ptr;
};
```

This could get hairy, because it might not see the `count` in `struct B`. BUT if it's not a pointer:

```
struct A {
  int count;
  struct B {
    int count;
    int arr[] __counted_by(count);
  } z;
};
```

Then I *do* want it to reference the `count` in `struct A`. Hilarious, I know! :-) The "pointer" example should work exactly as if `struct B` was defined outside of `struct A`. I'll need to make sure to document that very carefully.

As for other `-fbounds-safety` checks, I haven't seen any that do quite what we're doing here. When I looked through the attributes, few of them seemed to reference other fields within structures, or even external variables. (I know some exist, but they're not quite the same as what we're doing here.) Then again, I may quite well have not noticed one of the bounds safety code paths.

Reply:
I should say that I'm not *certain* that `ptr->arr` in the middle case will see `struct A` as its most-enclosing struct. I need to check on it.

Reply:
But *why* are you implementing a different rule here than we use for other uses of `__counted_by`?  That's an actively bad thing to do.

Look, if we have a use case that requires us to make `__counted_by` more flexible than it currently is, then okay, let's have a proper conversation about that and what the design ought to be.  The right place to do that is an RFC, not a code review thread.  Assuming we accept that proposal, though, this seems like the wrong way to implement it, because it's basically creating a bunch of ad hoc differences just for flexible array members.  There's no reason that `__counted_by` should follow different rules for flexible array members than it does for pointer members, at least on the read/bounds-checking side.  That means that we should be able to use common infrastructure between the two paths for getting back to the root pointer for the bound expression and evaluating it.  Your code here should just be invoking that common infrastructure.

So what I suggest is that you do this in three phases:
1. Make `__counted_by` work with flexible array members using the common infrastructure for `__counted_by`.
2. Get agreement on how we're going to generalize `__counted_by` and what the behavior ought to be.
3. Implement the new behavior for `__counted_by` in the common infrastructure for *all* its applications.

Reply:
What do you mean the other uses of `__counted_by`? It's only used with FAMs and in the sanitizer. This patch doesn't add new functionality. It only some bugs and modifies how the code is created.

Reply:
We have other code that computes the value of the `__counted_by` bound when we e.g. load a pointer from a field with the attribute, yes?

Reply:
No. there are only two places where we look at the `__counted_by` attribute: during bounds checking, and in a `__builtin_dynamic_object_size` statement. The former is the main focus of this patch. The latter attempts to calculate the size without introducing side-effects from the loading `count`. They both use similar paths, though I've been toying with your idea of using the offsets and it seems to work and may be more robust than trying to build the array and count accesses off of the same base pointer.

Reply:
I understand that this patch specifically can be thought of as trying to do something fairly modest.  However, it is a part of a sequence of patches that I haven't been paying much attention to, and what it is actually doing makes me concerned about the larger sequence of patches and whether you might be implementing something that doesn't make a whole lot of sense. I am trying to understand what's going on to see if maybe I can course-correct and make sure that the overall feature is a good one.

Reply:
Okay, I think I'm starting to see the root of my confusion here.  It looks like you *are* updating the common infrastructure for computing bounds. You are *also* adding a new bounds check for array subscripts into flexible array members, and that code seems to be quite a bit more complex than just calling into the common infrastructure.

@rapidsna Is there a specification for bound expressions?  I would like to understand how this works a little better, and in particular I am concerned about the idea that bound expressions can involve a path that isn't rooted in the same struct as the bounded member.

Reply:
Forcing the `count` to be in the top-level struct was more-or-less a compromise in how the `__counted_by` field is specified. It might be too restrictive, or not restrictive enough (depending on your point of view). I personally would prefer a different way of specifying where the `count` is, ですけど.... (My idea is to use a form of the "designated initializer" syntax (maybe with a way to specify a "relative path?" ... dunno).)

If supporting a FAM in a substructure is too hairy, I can remove that support until we work out the kinks. (I think GCC doesn't yet have support for this.)

Reply:
Well, like I said, I don't think there's a reason to use *different* rules for flexible array members one way or the other. But that's what this patch seems to be doing.

How did you end up going down the rabbit hole of writing all the new code for FAMs here?  If you were writing comprehensive tests for FAMs and finding that the substructure cases didn't work, I could believe that the real problem is that the common infrastructure doesn't really support that yet.  If that's the case, then yeah, I think the right path forward is to just worry about the cases that the common infrastructure supports, and we should make a point of going back to the design stage and trying to decide whether the current generality makes any sense.  A feature we can't actually implement doesn't do anyone any good.

Reply:
It started with a couple of bugs reported by someone testing the feature on Linux. I created a fix, but it wasn't super great and relied upon some assumptions about how the FAM expressions were formatted. Plus it could reload the expressions, which isn't good, especially if they have side effects. @nickdesaulniers had the good idea to generate the GEPs in a similar way to this. (That's where this code comes from.) Then several different cases were mentioned that needed to be handled at least gracefully...and the yak kept growing hair.

I don't necessarily think that this is adding "new" features more than just taking care of case after case after case ..., which is legion in C. For instance, the code was meant to handle FAMs in substructures, but because it's potentially very messy, maybe I should remove that feature. That would hopefully simplify this patch.

I'll update this patch removing the substructure FAM support and using your "offset" idea to get the count. Thanks for your feedback!

Reply:
@rjmccall @bwendling I have documentation clarifying what are the restrictions for __counted_by https://github.com/rapidsna/llvm-project/pull/1/commits/4a6a041267b3399eab46c3bb88e3d023d4050836. I was planning to land it after my initial documentation lands https://github.com/llvm/llvm-project/pull/70749.

@bwendling In general, the same name/expression rules should apply to `__counted_by` used in pointers and arrays. I clarified that in https://github.com/rapidsna/llvm-project/pull/1/commits/4a6a041267b3399eab46c3bb88e3d023d4050836. I synched with GCC folks and they agreed that the syntax and semantics should not diverge between the full `-fbounds-safety` mode and the mode only using the annotations for `__bdos` and sanitizers.

Reply:
Thanks, @rapidsna. I agree with what you said here. (I also agree with @rjmccall.) I don't want to make anymore changes to this patch though. As @nathanchance pointed out, this is holding up Linux testing / development and I want to submit it soon. I believe it's ready for submission.

To be honest, this patch was only meant as a bug fix, but snowballed out of all proportions. However, I think it's in a state where we could submit this and continue development from there.

I need to point out that this *isn't* the last patch in this feature. I'll have more to follow. Can someone please just give approval so that we can get on with life?